Lib/test/test_remote_pdb.py

```diff
@@ -78,6 +79,697 @@ def get_output(self) -> List[dict]:
         return results
 
 
+class MockDebuggerSocket:
+    """Mock file-like simulating a connection to a _RemotePdb instance"""
+
+    def __init__(self, incoming):
+        self.incoming = iter(incoming)
+        self.outgoing = []
+        self.buffered = bytearray()
+
+    def write(self, data: bytes) -> None:
+        """Simulate write to socket."""
+        self.buffered += data
+
+    def flush(self) -> None:
+        """Ensure each line is valid JSON."""
+        lines = self.buffered.splitlines(keepends=True)
+        self.buffered.clear()
+        for line in lines:
+            assert line.endswith(b"\n")
+            self.outgoing.append(json.loads(line))
+
+    def readline(self) -> bytes:
+        """Read a line from the prepared input queue."""
+        # Anything written must be flushed before trying to read,
+        # since the read will be dependent upon the last write.
+        assert not self.buffered
+        try:
+            item = next(self.incoming)
+            if not isinstance(item, bytes):
+                item = json.dumps(item).encode()
+            return item + b"\n"
+        except StopIteration:
+            return b""
+
+    def close(self) -> None:
+        """No-op close implementation."""
+        pass
+
+
+class PdbClientTestCase(unittest.TestCase):
+    """Tests for the _PdbClient class."""
+
+    def do_test(
+        self,
+        *,
+        incoming,
+        simulate_write_failure=False,
+        simulate_flush_failure=False,
+        expected_outgoing=[],
```

Comment:
Let's make this and `expected_completions` default to `None` - and if they are `None`, we don't check it (instead of checking that it was empty). That gives us the flexibility to ignore some of the results. Also I would avoid having a mutable object as a default value (even though it might happen to work).

Reply:
I don't think that's reasonable. The whole point of these tests is to make assertions about how the client interacts with the server. Any test that makes no assertions about what the client sends to the server is a bad test.

Reply:
We can keep it as it is. If in the future we need that flexibility we can do it. It might not be necessary when we are dealing with a mock socket, but it might be for real integration tests.